src/System.Drawing.Common/tests/IconTests.cs

```diff
@@ -0,0 +1,586 @@
+using System.Runtime.Serialization.Formatters.Binary;
+using Xunit;
+
+namespace System.Drawing.Tests
+{
+    public class IconTests
+    {
+        [Theory]
+        [InlineData("smiley.ico")]
+        [InlineData("323511.ico")]
+        public void Ctor_FilePath(string name)
+        {
+            var icon = new Icon(Helpers.GetEmbeddedResourcePath("smiley.ico"));
+            Assert.Equal(32, icon.Width);
+            Assert.Equal(32, icon.Height);
+            Assert.Equal(new Size(32, 32), icon.Size);
+        }
+
+        public static IEnumerable<object[]> Size_TestData()
+        {
+            // Normal size
+            yield return new object[] { "smiley.ico", new Size(16, 16), new Size(16, 16) };
+            yield return new object[] { "smiley.ico", new Size(-32, -32), new Size(16, 16) };
+            yield return new object[] { "smiley.ico", new Size(32, 16), new Size(32, 32) };
+            yield return new object[] { "323511.ico", new Size(48, 48), new Size(48, 48) };
+            yield return new object[] { "323511.ico", new Size(0, 0), new Size(32, 32) };
+            yield return new object[] { "323511.ico", new Size(1, 1), new Size(256, 256) };
+
+            // Unusual size
+            yield return new object[] { "80509.ico", new Size(16, 16), new Size(10, 16) };
+            yield return new object[] { "80509.ico", new Size(32, 32), new Size(11, 22) };
+
+            // Only 256
+            yield return new object[] { "only256.ico", new Size(0, 0), new Size(256, 256) };
+        }
+
+        [Theory]
+        [MemberData(nameof(Size_TestData))]
+        public void Ctor_FilePath_Width_Height(string fileName, Size size, Size expectedSize)
+        {
+            var icon = new Icon(Helpers.GetEmbeddedResourcePath(fileName), size.Width, size.Height);
+            Assert.Equal(expectedSize.Width, icon.Width);
+            Assert.Equal(expectedSize.Height, icon.Height);
+            Assert.Equal(expectedSize, icon.Size);
+        }
+
+        [Theory]
+        [MemberData(nameof(Size_TestData))]
+        public void Ctor_FilePath_Size(string fileName, Size size, Size expectedSize)
+        {
+            var icon = new Icon(Helpers.GetEmbeddedResourcePath(fileName), size);
+            Assert.Equal(expectedSize.Width, icon.Width);
+            Assert.Equal(expectedSize.Height, icon.Height);
+            Assert.Equal(expectedSize, icon.Size);
+        }
+
+        [Fact]
+        public void Ctor_NullFilePath_ThrowsArgumentNullException()
+        {
+            Assert.Throws<ArgumentNullException>("path", () => new Icon((string)null));
+            Assert.Throws<ArgumentNullException>("path", () => new Icon((string)null, new Size(32, 32)));
+            Assert.Throws<ArgumentNullException>("path", () => new Icon((string)null, 32, 32));
+        }
+
+        [Fact]
+        public void Ctor_Stream()
+        {
+            using (var stream = new FileStream(Helpers.GetEmbeddedResourcePath("smiley.ico"), FileMode.Open))
```

Comment:
This could be using `Assembly.GetManifestResourceStream`, though perhaps these files should just not be embedded resources in the first place.